### PR TITLE
Selections and Producers

### PR DESCRIPTION
**Pull request by kreczko**
For this PR I am planning to
 - [x] Add new selection interface
 - [x] Add InvariantMass producer

The current implementation is not yet fully working, even if I use an empty selection:
```python
selection = Selection(
        dict(
            All=()
        )
    )

selection.set_monitoring_file('test.txt')
```
The code:
```python
dataframes = tree.scan(tblcfg=tables, max_events=50, selection=selection)

    for df in dataframes:
        print df
        print df.to_string(index=False)
```
Produces the error
```
[('component', 'depth', 'class', 'name', 'pass', 'total')]
Traceback (most recent call last):
  File "run.py", line 57, in <module>
    print df.to_string(index=False)
AttributeError: 'list' object has no attribute 'to_string'
```

If somebody can think of an easy unit test for this problem that does not use data, please let me know.

### CODE REVIEW

**Comment by kreczko** [comment]
This PR is now ready for review.

Invariant mass producer example output for Z candidates:
di_muon_mass

**Comment by TaiSakuma** [comment]
It is good have test.

But I am not sure if nose is a good choice at the moment.

I thought about using nose for other projects. But then I didn't use it because [its website](http://nose.readthedocs.io/en/latest/) says that "Nose has been in maintenance mode for the past several years and will likely cease without a new person/team to take over maintainership. New projects should consider using Nose2, py.test, or just plain unittest/unittest2."

Shouldn't we consider using other test packages as the nose website suggests?

**Comment by kreczko** [comment]
Ah, yes. It has been on my list for a while, but I am slipping back into old habits.

The `nose2` syntax is close to Ruby and seems to be very easy to construct story-like tests:
http://nose2.readthedocs.io/en/latest/such_dsl.html

**Comment by TaiSakuma** [comment]
ok. maybe we should switch to nose2 or pytest at some point.

File: test/test_selection.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    )
+
+    expected = dict(
+        All=(
+            'ev: ev.Muon_E[0] > 2 * ev.Muon_Pt[1]',
+            'ev: ev.Muon_Charge[0] == ev.Muon_Charge[0]',
+        )
+    )
+
+    assert_equals(selection._convert_to_alphatwirl(), expected)
```
[the remark this answers is not in the document]
Comment (kreczko) [reply test/test_selection.py]: The translation rules are defined in [selection.py#L24](https://github.com/kreczko/alphatwirl-interface/blob/cc651ee84d6cbb0ddb4ad02341ade30634a64c11/alphatwirl_interface/selection.py#L24)
it uses regular expressions to figure out the variables/branches used in the string and inserts a token, `ev.` in front of them and then makes it into a lambda string.
so
 -  `'NMuon == 2'` &rarr; `ev: ev.NMuon == 2'`
 - `'Muon_Pt[0] > 20'` &rarr; 'Muon_Pt[0] > 20'
However, it is currently an incomplete translater as it will mistranslate `and`, `or` and functions (e.g. `abs`).

TL;DR
Mapping is flexible and can be adjusted.

[the remark this answers is not in the document]
Comment (TaiSakuma) [reply test/test_selection.py]: Thank you for the reply.

In general, I don't really agree that we pursue this direction. In this direction, we will basically end up developing our own pseudo language that works only as arguments of methods of Selection.

If we don't manipulate the string, it is clear what it means. For example, `ev: ev.NMuon == 2` will be executed as a python code, more precisely will be an argument to lambda written as a string. This is the complete description of how the string works. If we write this description in documentations, users will have the complete knowledge of how this string functions. 

In contrast, what is 'NMuon == 2'? What language is this? This is not written in any language. The string will be manipulated first and then will be given to lambda.

I know the regular expression is used. My question was what are the intended manipulation?
- add `ev: ` to the beginning
- add `ev.` in front of the names of the attributes of the object `ev`
- add `[0]` after names of the attributes of the object `ev`
Are these 3 the complete list of the manipulation? 

As you already pointed out, there will be mistranslation. `and`, `or`,  the functions are not complete list of the cases in which mistranslations occur.
- Should `True` be translated into `ev: True` or `ev: ev.True`?
- The current regex doesn't match a single letter. Is it intentional?
- How about `1 in (bintypeId, bintypeIdJECUp, bintypeIdJECDown)`? `in` is going to be matched? is it intentional?
- How about `primaryDataset == "MET"`? The quoted string `"MET"` will be matched and turned into `ev."MET"`.

I can continue the list for a while. There will also be many cases I cannot think of now.

It will be a lot work to make this manipulation work perfectly or even reasonably well. It will be as lot work as developing new little language.

I wouldn't manipulate this string and would just give it to lambda.
